### DISCUSSION TITLE
error loading a custom extension after upgrading to v0.10.2 from v0.9.2

### DISCUSSION THREAD

**Discussion by softprops** [duckdb/duckdb#12000] [General]
I'm hitting an error reported as `Invalid Input Error: Extension "./zig-out/lib/quack.duckdb_extension" do not have metadata compatible with DuckDB loading it (version v0.10.2, platform osx_arm64)` 
around [these lines](https://github.com/duckdb/duckdb/blob/d9efdd14270245c4369096e909acecea174d86cc/src/main/extension/extension_load.cpp#L190-L198) when I run `duckdb -unsigned -s "LOAD './zig-out/lib/quack.duckdb_extension';"`

I am about to build and load the same extension in 0.9.2 without issue

I just upgrade my duckdb cli to 0.12.2

```sh
duckdb --version
v0.10.2 1601d94f94
```

I'm building against [`v0.10.2` tag of duckdb](https://github.com/duckdb/duckdb/tree/v0.10.2) (hash `1601d94f94`) the version which lines up with the sha I see output from `duckdb --version`. I'm returning is `v0.10.2` which is the result of the c api bindings `duckdb_library_version()` function. This led me to these [3 lines](https://github.com/duckdb/duckdb/blob/d9efdd14270245c4369096e909acecea174d86cc/src/main/extension/extension_load.cpp#L190-L192) which I don't fully understand what they represent, it looks to be a hardcoded magic number 4 but I'm not sure where that comes form to how to align with it.

I'm looking for insight on what I might be missing. Can anyone here help?

**Comment by softprops** [comment]
I believe this relates to https://github.com/duckdb/duckdb/pull/11515 

from the diff I've been able to derive an incantation to avoid this issue but it seems less than ideal

```sh
set allow_extensions_metadata_mismatch = true
LOAD './zig-out/lib/quack.duckdb_extension';
```

One needs to understand both c++ and cmake as each our their own respective languages to understand this pull and I know neither. I'm building with zig but I believe the key changes somewhere within cmake in this pull. 

Is the protocol for what duckdb expects documented somewhere as part of a formal specification?

**Reply by carlopi** [reply]
Can you share a repo ? I am curious to see whether it's possible to recycle the current cmake logic, otherwise I will provide a Python script.

I have yet to properly write document how to make this work, but basics are:

### hack mode
fetch a duckdb extension for your given platform
truncate the last 512 bytes
append those 512 bytes to your own extension
this should now work

### explanation
there are currently 512 bytes that needs to be appended at the end of extensions.
256 bytes of metadata, and 256 bytes of signature.
Metadata is currently represented by 8 fields of 32 bytes.
Currentlty, field 1-4 are not considered, are a future extension.
Field 4-8 are currently read, and they are: 32 bytes for extension version, 32 bytes for duckdb version, 32 bytes for platform, 32 bytes for magic bytes encoding the metadata version.
Metadata: platform, version and metadata version are read and tested against
Signature: for unsigned extensions, the signature bytes are not really relevant, but they have to be there (even 256 `0x00` bytes work)

### Proper solution
We (= me in this case) need to provide a script appending the relevant values (I have it already as Python script, but need to check it actually works as intended)

**Reply by softprops** [reply]
thanks! a python script would be great. 

here's what Im working with atm https://github.com/softprops/zig-duckdb-ext. I'm writing a library for zig extensions and was then going to provide a separate repo as a template that consumes it. in its current state this repo is both. 

I'm trying to work my way backwards from 

* https://github.com/duckdb/duckdb/blob/d9efdd14270245c4369096e909acecea174d86cc/src/main/extension/extension_load.cpp#L156-L180
* https://github.com/duckdb/duckdb/blob/main/scripts/append_metadata.cmake 
* and the pull description

I think what you outlined above is what I'm approximating this as so far so I think Im on the right track

If you could share that python script I could test it against what I have

I understand what you're trying to do in spirit. in 0.9.2 you have to load the library to derive the version. With recent changes you don't need to load it to validate that its compatible. That makes sense to me. Does that mean extensions technically no longer. have to export an [`{extension-name}_version()`](https://github.com/softprops/zig-duckdb-ext/blob/81ff3c25300cc9331f4c6132aadc1cf0565bc067/src/bridge.cpp#L20-L24) function? If it's in the metadata, duckdb will already know the version.

**Comment by carlopi** [comment]
> thanks! a python script would be great.
> 
> here's what Im working with atm https://github.com/softprops/zig-duckdb-ext. I'm writing a library for zig extensions and was then going to provide a separate repo as a template that consumes it. in its current state this repo is both.

Will take a look on how to best integrate this, but very basic that will work is:
```
echo -n "my_platform" > duckdb_platform_out
#my_platform needs to be the same as the output of ./duckdb -c "PRAGMA platform";
touch extension_file
cmake -DEXTENSION=extension_file -DDUCKDB_VERSION="v0.10.2" -DEXTENSION_VERSION="my_extension_version" -DNULL_FILE=path/to/duckdb/scripts/null.txt -P path/to/duckdb/scripts/append_metadata.cmake
hexdump -c extension_file
00000000  00 93 04 10 64 75 63 6b  64 62 5f 73 69 67 6e 61  |....duckdb_signa|
00000010  74 75 72 65 80 04 00 00  00 00 00 00 00 00 00 00  |ture............|
00000020  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
00000090  00 00 00 00 00 00 6d 79  5f 65 78 74 65 6e 73 69  |......my_extensi|
000000a0  6f 6e 5f 76 65 72 73 69  6f 6e 00 00 00 00 00 00  |on_version......|
000000b0  00 00 00 00 00 00 76 30  2e 31 30 2e 32 00 00 00  |......v0.10.2...|
000000c0  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
000000f0  00 00 00 00 00 00 34 00  00 00 00 00 00 00 00 00  |......4.........|
00000100  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
00000210  00 00 00 00 00 00                                 |......|
00000216
```

Super cool, will take a look at the repo later.

> I understand what you're trying to do in spirit. in 0.9.2 you have to load the library to derive the version. With recent changes you don't need to load it to validate that its compatible. That makes sense to me. Does that mean extensions technically no longer. have to export an [`{extension-name}_version()`](https://github.com/softprops/zig-duckdb-ext/blob/81ff3c25300cc9331f4c6132aadc1cf0565bc067/src/bridge.cpp#L20-L24) function? If it's in the metadata, duckdb will already know the version.

Yep, altough this might have sailed in the sense that if you open your extension with older duckdb version it will still expect to find the version. Arguably that is broken already either way, but we haven't made a call on properly removing this due to this.
But currently the only usefulness is in throwing a nicer error message when mistakenly opening this from an older duckdb (pre 0.10.2)

**Reply by softprops** [reply]
no worries. its late for me so calling it here. but wanted to confirm your hackmode worked!

https://github.com/softprops/zig-duckdb-ext/blob/4ce625a0603e21c4795549de1fccb79d3c4f0c85/Justfile#L27-L29

tldr

```
tail -c 512 ~/.duckdb/extensions/v0.10.2/osx_arm64/motherduck.duckdb_extension > metadata-bytes
cat metadata-bytes >> zig-out/lib/quack.duckdb_extension 
```

**Reply by softprops** [reply]
a few additional questions

I'm comparing what I took from the metadata payload of motherduck (it works) as a sample and am trying to reimplement this in zig

here's my motherduck sample

```sh
 hexdump -C metadata-bytes
00000000  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
000000a0  76 30 2e 31 30 2e 32 00  00 00 00 00 00 00 00 00  |v0.10.2.........|
000000b0  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
000000c0  6f 73 78 5f 61 72 6d 36  34 00 00 00 00 00 00 00  |osx_arm64.......|
000000d0  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
000000e0  34 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |4...............|
000000f0  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
00000100  b7 8a 06 f7 0c d2 57 01  38 4a b0 79 33 8e 78 f1  |......W.8J.y3.x.|
00000110  84 5a db 49 5d 69 8c d8  2b d2 cf 34 4b b2 67 5c  |.Z.I]i..+..4K.g\|
00000120  f6 35 86 da e9 cb fa 38  01 27 b9 4c 1e 5d ad e0  |.5.....8.'.L.]..|
00000130  2b 70 02 6e 40 94 dc 15  5f 22 11 e4 16 b6 89 4f  |+p.n@..._".....O|
00000140  11 ad 6c a2 44 8a 1d 3b  c7 c7 4e 37 83 26 92 a4  |..l.D..;..N7.&..|
00000150  ec a0 16 53 18 da a3 7f  13 21 44 eb 9f 13 6b cd  |...S.....!D...k.|
00000160  25 ca bd 79 77 0f b8 4b  1d 3b 02 67 21 ff e8 dd  |%..yw..K.;.g!...|
00000170  28 8a b8 96 ff 71 78 28  21 5c 27 e3 97 db e8 e7  |(....qx(!\'.....|
00000180  66 00 43 0d 2c ef 79 c0  fe c0 a0 6e 23 6e fc 12  |f.C.,.y....n#n..|
00000190  b7 9f 9e e6 26 9a 87 ef  a8 d3 54 87 44 5c 31 e4  |....&.....T.D\1.|
000001a0  f2 93 5f 1e d5 f8 dd e7  2b e0 c2 75 b2 e6 e3 be  |.._.....+..u....|
000001b0  1c 6a 93 53 ec d1 50 7f  e3 32 65 7e 56 38 b1 37  |.j.S..P..2e~V8.7|
000001c0  51 83 a6 38 90 e2 30 02  b5 a1 d2 d5 bd b9 02 f1  |Q..8..0.........|
000001d0  74 b2 57 01 f0 98 9b 1d  0b be 97 ba a2 4e d6 75  |t.W..........N.u|
000001e0  14 19 e3 b9 88 46 23 19  43 29 6d 36 d7 8a 79 dd  |.....F#.C)m6..y.|
000001f0  17 ef be 3e 65 5c 0d 89  29 aa 4d f9 38 6f fe 0c  |...>e\..).M.8o..|
```

I can see the metadata at the top and signature bytes at the bottom but I'm a big confused about how the metadata fields line up


> 256 bytes of metadata, and 256 bytes of signature.
Metadata is currently represented by 8 fields of 32 bytes.
Currentlty, field 1-4 are not considered, are a future extension.
Field 4-8 are currently read, and they are: 32 bytes for extension version, 32 bytes for duckdb version, 32 bytes for platform, 32 bytes for magic bytes encoding the metadata version.
Metadata: platform, version and metadata version are read and tested against
Signature: for unsigned extensions, the signature bytes are not really relevant, but they have to be there (even 256 0x00 bytes work)

the metadata fields actually look like they are in the reverse [order of those specified in the pull request](https://github.com/duckdb/duckdb/pull/11515) but do line up with what you mentioned. I think the fields are [reversed back](https://github.com/duckdb/duckdb/blob/d9efdd14270245c4369096e909acecea174d86cc/src/main/extension/extension_load.cpp#L182C15-L182C29) to the order of the pr when they are loaded, again my c++ is weak so I could be wrong. 

here's what I have so far


```zig
pub fn main() !void {
    var args = std.process.args();
    _ = args.next();

    const path = args.next() orelse {
        @panic("expected path argument. this is the path to the file we're appending metadata to");
    };
    const platform = args.next() orelse {
        @panic("expected duckdb platform argument. see https://duckdb.org/docs/extensions/working_with_extensions#platforms");
    };
    const ddbVersion = args.next() orelse {
        @panic("expected duckdb version argument. this should be set to the version this ext was linked against");
    };
    const extVersion = args.next() orelse ddbVersion;
    std.debug.print(
        "\nappending metadata to path {s}\n * platform {s}\n * ddb version {s}\n * ext version {s}\n",
        .{
            path,
            platform,
            ddbVersion,
            extVersion,
        },
    );
   // a 512 block zero'd bytes
    var payload = std.mem.zeroes([512]u8);
   //  fields are in reversed order they will be processed by ext loading
    const segments = [_][]const u8{
        extVersion,
        ddbVersion,
        platform,
        "4",
    };
    for (segments, 0..) |segment, i| {
        const start = 32 * i;
        const end = start + segments[i].len;
        @memcpy(payload[start..end], segment);
    }
    var file = try std.fs.cwd().openFile(path, .{ .mode = .read_write });
    try file.seekTo(try file.getEndPos());
    try file.writer().writeAll(&payload);
}
```

here's a hexdump output of what this generates. 

```sh
hexdump -C test-metadata 
00000000  76 30 2e 31 30 2e 32 00  00 00 00 00 00 00 00 00  |v0.10.2.........|
00000010  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
00000020  76 30 2e 31 30 2e 32 00  00 00 00 00 00 00 00 00  |v0.10.2.........|
00000030  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
00000040  6f 73 78 5f 61 72 6d 36  34 00 00 00 00 00 00 00  |osx_arm64.......|
00000050  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
00000060  34 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |4...............|
00000070  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
00000200
```

note when I compare this to the motherduck sample metadata type my field data starts at each index of the 32 slots. it kind of looks like in the mother duck example they are aligned in some other within that 32 byte block

I think I'm getting close but still not quite there.

**Reply by softprops** [reply]
update, I got it I was "half" the way there. there are actually EIGHT slots in reverse order they are processed starting with 4 empty stlots. I was only inserting 4.

the following worked for me 

```zig
const std = @import("std");

pub fn main() !void {
    var args = std.process.args();
    _ = args.next();

    const path = args.next() orelse {
        @panic("expected path argument. this is the path to the file we're appending metadata to");
    };
    const platform = args.next() orelse {
        @panic("expected duckdb platform argument. see https://duckdb.org/docs/extensions/working_with_extensions#platforms");
    };
    const ddbVersion = args.next() orelse {
        @panic("expected duckdb version argument. this should be set to the version this ext was linked against");
    };
    const extVersion = args.next() orelse ddbVersion;
    std.debug.print(
        "\nappending metadata to path {s}\n * platform {s}\n * ddb version {s}\n * ext version {s}\n",
        .{
            path,
            platform,
            ddbVersion,
            extVersion,
        },
    );
    var payload = std.mem.zeroes([512]u8);
    const segments = [_][]const u8{
        "",         "",         "",       "", // <= 👈 reserved but unused metadata slots
        extVersion, ddbVersion, platform, "4",
    };
    for (segments, 0..) |segment, i| {
        const start = 32 * i;
        const end = start + segments[i].len;
        @memcpy(payload[start..end], segment);
    }
    var file = try std.fs.cwd().openFile(path, .{ .mode = .read_write });
    try file.seekTo(try file.getEndPos());
    try file.writer().writeAll(&payload);
}
```

note how the prefix now looks a lot more like the motherduck example.

```sh
hexdump -C test-metadata                                                                                                                              5s
00000000  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
00000080  76 30 2e 31 30 2e 32 00  00 00 00 00 00 00 00 00  |v0.10.2.........|
00000090  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
000000a0  76 30 2e 31 30 2e 32 00  00 00 00 00 00 00 00 00  |v0.10.2.........|
000000b0  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
000000c0  6f 73 78 5f 61 72 6d 36  34 00 00 00 00 00 00 00  |osx_arm64.......|
000000d0  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
000000e0  34 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |4...............|
000000f0  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
00000200
```
